packages/styled/types/tests.tsx

```diff
@@ -217,3 +217,69 @@ const Input5 = styled.input`
   // $ExpectError
   ;<StyledCompWithoutAs as={Section} />
 }
+
+{
+  // Props forwarding through StyledOptions and FilteringStyledOptions
+
+  const fc: React.FC<{ foo: string }> = props => <div {...props} />
+
+  styled(fc, { shouldForwardProp: (prop: 'foo') => true })({})
```

Comment:
this is somewhat interesting - `shouldForwardProp` here gets typed through this:
```
StyledOptions<React.PropsWithChildren<{ foo: string; }>>.shouldForwardProp?(propName: "foo" | "children"): boolean
```
So I'm unsure why does this typecheck if this should be an error:
[TS playground](https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEAzArgOzAFwJYHtXygAooAueAckRx3PgB8KwALLCYOVcgSjICNqIIKKgBQI0JFgIU6bHni9iZStW58BQ0ePDQ48MHgDOGfWQwBPAA4gciBWIOpj8YGas27UeAF59QA)

Reply:
Hmm... I think It depends on the direction of comparison. [Type swap fixes the example](https://www.typescriptlang.org/play?ssl=5&ssc=22&pln=1&pc=1#code/CYUwxgNghgTiAEAzArgOzAFwJYHtXygAooAueAckRx3PgB8KwALLCYOVcgSjICNqIIKKgBQoSLAQp02PPF7Eylatz4ChoseGhx4YPAGcMeshgCeABxA5EBEftRH4wU5eu3e8ALx6gA)

```ts
declare function a(a: 'foo' | 'children'): boolean
declare function b(a: 'foo'): boolean

declare const c: typeof a
const d: typeof b = c
```

Reply:
So this comes from the fact that `shouldForwardProp` is currently declared using "object method" syntax, and not using the property syntax (where the value turns out to be a function). This makes the function bivariant instead of contravariant (I've expected the latter here).

Take a look at this example that might showcase the difference:
[TS playground](https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEAzArgOzAFwJYHtXwGcAHcLRLMAMSwgxBgAoBrEATwC54AiRHHLgJScARnwggoqANwAoGaEiwEKdNjzwoYMCCIYAwngwwoAN1hZJGBjl0FOAbxnwkNOjE7M2nAkayoA5gLwALwAfPCiOOKSsgC+QvAmOFjAsjKa2roGqEam5pYMjs7ktPTeJGBkFNSlMAA0MvFpCtBwSGiYuPgZOhgAQlhmMBY51rYOLrWeHIS+AQmR0fjxnEkpaT26A0MjVkWTbuWk5FSu9A1NMkA)

I feel like usually bivariant functions are not the ones that we want so it's a good time to fix that old mistake with this patch:
```diff
diff --git a/packages/styled/types/base.d.ts b/packages/styled/types/base.d.ts
index 4c972110..83949462 100644
--- a/packages/styled/types/base.d.ts
+++ b/packages/styled/types/base.d.ts
@@ -23,15 +23,15 @@ export interface FilteringStyledOptions<
   ForwardedProps extends keyof Props = keyof Props
 > {
   label?: string
-  shouldForwardProp?(propName: PropertyKey): propName is ForwardedProps
+  shouldForwardProp?: (propName: PropertyKey) => propName is ForwardedProps
   target?: string
 }
 
 export interface StyledOptions<Props = null> {
   label?: string
-  shouldForwardProp?(
+  shouldForwardProp?: (
     propName: Props extends null ? PropertyKey : keyof Props
-  ): boolean
+  ) => boolean
   target?: string
 }
```

Note that this might require some tests to be adjusted

Reply:
It does seem like this line of code 

```ts
styled(fc, { shouldForwardProp: (prop: 'foo') => true })({})
```

should error. You should not be allowed to pass in a `shouldForwardProp` that does not handle all props of the component.

It sounds like @Andarist's patch will fix this.